### ISSUE TITLE
not calling primitive function correctly

### ISSUE THREAD

**Issue by timsgardner** [nasser/magic#15]
clojure:
```clojure
(mc/define lookahead-acc ^UnityEngine.Vector3 [gobj
                                               ^UnityEngine.Vector3 p1,
                                               ^UnityEngine.Vector3 p2,
                                               ^UnityEngine.Vector3 vel,
                                               ^System.Single max-accel]
  (let [dist (UnityEngine.Vector3/Distance p1 p2)
        brake-time (* 2.0
                      (/ (UnityEngine.Mathf/Abs (.magnitude vel))
                         max-accel))
        lookahead (v3+ p1 (v3* vel brake-time))
        lookahead-diff (v3- p2 lookahead)
        retval (v3* (normalized lookahead-diff)
                    (min max-accel (magnitude lookahead-diff)))]
    retval))

(defn go-to [^GameObject obj]
  (when-let [target (state obj ::target)]
    (let [max-accel (float (or (state obj ::max-accel)
                               (throw
                                 (Exception.
                                   "Expects :car-scene.movement/max-accel key in state"))))]
      (with-cmpt target [targ-tr UnityEngine.Transform]
        (with-cmpt obj [tr UnityEngine.Transform
                        rb UnityEngine.Rigidbody]
          (mf/faster
            (let [vel (.velocity rb)
                  difvec (v3- (.position targ-tr) (.position tr))
                  pos (.position tr)
                  targ-pos (.position targ-tr)]
              (if (or (< 0.001 (UnityEngine.Mathf/Abs (UnityEngine.Vector3/Magnitude vel)))
                      (< 0.001 (UnityEngine.Vector3/Distance pos targ-pos)))
                (let [force-vec (v3scale (lookahead-acc obj,  pos, targ-pos, vel, max-accel)
                                         (v3 1))]
                  (.AddForce rb force-vec UnityEngine.ForceMode/Acceleration))
                (sets! rb velocity (v3 0))))))))))
```
Disassembly:
```C#
using clojure.lang;
using System;
using UnityEngine;

public class magic$$car-scene$movement$$lookahead-acc$$379 : AFunction, Function<Vector3, object, Vector3, Vector3, Vector3, float>, Function<object, object, object, object, object, object>
{
	//
	// Static Fields
	//
	private static readonly Var car-scene$movement/magnitude;

	private static readonly Var car-scene$movement/normalized;

	//
	// Constructors
	//
	static magic$$car-scene$movement$$lookahead-acc$$379 ()
	{
		magic$$car-scene$movement$$lookahead-acc$$379.car-scene$movement/magnitude = RT.var ("car-scene.movement", "magnitude");
		magic$$car-scene$movement$$lookahead-acc$$379.car-scene$movement/normalized = RT.var ("car-scene.movement", "normalized");
	}

	//
	// Methods
	//
	public override bool HasArity (int num)
	{
		return num == 5;
	}

	public override Vector3 invoke (object obj, Vector3 vector, Vector3 vector2, Vector3 vector3, float num)
	{
		float num2 = Vector3.Distance (vector, vector2);
		double arg_2B_0 = 2.0;
		Vector3 vector4 = vector3;
		double num3 = Numbers.multiply (arg_2B_0, Numbers.divide ((double)Mathf.Abs (vector4.get_magnitude ()), (double)num));
		float num4 = RT.floatCast (num3);
		Vector3 vector5 = vector + vector3 * num4;
		Vector3 vector6 = vector2 - vector5;
		float num5 = RT.floatCast (Numbers.min ((double)num, (double)((Function<float, Vector3>)magic$$car-scene$movement$$lookahead-acc$$379.car-scene$movement/magnitude.getRawRoot ()).invoke (vector6)));
		return ((Function<Vector3, Vector3>)magic$$car-scene$movement$$lookahead-acc$$379.car-scene$movement/normalized.getRawRoot ()).invoke (vector6) * num5;
	}

	public override object invoke (object obj, object obj2, object obj3, object obj4, object obj5)
	{
		return this.invoke (obj, (Vector3)obj2, (Vector3)obj3, (Vector3)obj4, RT.floatCast (obj5));
	}
}

```
```C#
using clojure.lang;
using System;
using UnityEngine;

public class magic$$car-scene$movement$$--faster--body402--$$405
{
	//
	// Static Fields
	//
	private static readonly Var car-scene$movement/lookahead-acc;

	//
	// Constructors
	//
	static magic$$car-scene$movement$$--faster--body402--$$405 ()
	{
		magic$$car-scene$movement$$--faster--body402--$$405.car-scene$movement/lookahead-acc = RT.var ("car-scene.movement", "lookahead-acc");
	}

	//
	// Static Methods
	//
	public static Rigidbody invoke (object obj, GameObject gameObject, float num, GameObject gameObject2, Transform transform, object obj2, Rigidbody rigidbody, Transform transform2, object obj3, GameObject gameObject3)
	{
		Vector3 velocity = rigidbody.get_velocity ();
		Vector3 vector = transform2.get_position () - transform.get_position ();
		Vector3 position = transform.get_position ();
		Vector3 position2 = transform2.get_position ();
		bool flag = Numbers.lt (0.001, (double)Mathf.Abs (Vector3.Magnitude (velocity)));
		Rigidbody arg_10B_0;
		if (!(flag ? flag : Numbers.lt (0.001, (double)Vector3.Distance (position, position2)))) {
			long num2 = 0L;
			Vector3 velocity2 = new Vector3 ((float)num2, (float)num2, (float)num2);
			rigidbody.set_velocity (velocity2);
			arg_10B_0 = rigidbody;
		}
		else {
// BUG:
			Vector3 arg_F4_0 = (Vector3)((IFn)magic$$car-scene$movement$$--faster--body402--$$405.car-scene$movement/lookahead-acc.getRawRoot ()).invoke (gameObject, position, position2, velocity, num);
			long num3 = 1L;
			Vector3 vector2 = Vector3.Scale (arg_F4_0, new Vector3 ((float)num3, (float)num3, (float)num3));
			rigidbody.AddForce (vector2, 5);
			arg_10B_0 = null;
		}
		return arg_10B_0;
	}
}
```
Above, in `((IFn)magic$$car-scene$movement$$--faster--body402--$$405.car-scene$movement/lookahead-acc.getRawRoot ()).invoke (gameObject, position, position2, velocity, num);`, uses `IFn` rather than `Function<Vector3, object, Vector3, Vector3, Vector3, float>`

**Comment by timsgardner** [comment]
Problem seems to go away when I get rid of the `obj` argument to `lookahead-acc`. Do generic `Function`s require all the arguments to be value-types?

**Comment by nasser** [comment]
There is no requirement by design that all arguments be value types, no. It might be looking for *exact* matches, though. I notice that you didn't typehint `gobj` in `lookahead-acc`, and it takes the type `object`, which is not the exact same type as `GameObject`. If you type hint that parameter with `GameObject` does the problem go away?

**Comment by nasser** [comment]
Confirmed it's an exact match issue

```clojure
(magic/define foo [^Single a] a)
(magic/define bar [] (foo 5.5)) ;; IFn

(magic/define foo [^Single a] a)
(magic/define bar [] (foo (float 5.5))) ;; Function<...>

(magic/define foo [^Double a] a)
(magic/define bar [] (foo 5.5)) ;; Function<...>
```

The presence of *any* parameter that is not an exact match will fall back to `IFn`. This is a bug.

**Comment by nasser** [comment]
Fixed in e16a644131a1ad2a9fc96e5d8e0232e74ef33b69
